### DISCUSSION TITLE
Relationship using column from third table

### DISCUSSION THREAD

**Discussion by Basskip** [sqlalchemy/sqlalchemy#8202] [Usage Questions] [answered]
Not very experienced with SQL or SQLAlchemy, but I'm trying to set up a relationship that is basically filtering an existing relationship by information that's available on a third table. 

I created a stripped down example of what I'm trying to do, have something like math_result on Student, which requires information from both Subject and SubjectResult

```
from sqlalchemy import create_engine, Column, Integer, String, ForeignKey
from sqlalchemy.orm import relationship, sessionmaker, selectinload
from sqlalchemy.ext.declarative import declarative_base

engine = create_engine('sqlite:///:memory:', echo=True)
Base = declarative_base()

class Student(Base):
    __tablename__ = 'students'
    id = Column(Integer, primary_key=True)
    name = Column(String)
    results = relationship(
        "SubjectResult",
        back_populates="student"
    )
    math_result = relationship(
        "SubjectResult",
        secondary="join(SubjectResult, Subject, and_(SubjectResult.subject_id == Subject.id, Subject.name == 'Math'))",
    )

    def __repr__(self):
        return f"<Student {self.id} {self.name}>"

class Subject(Base):
    __tablename__ = 'subjects'
    id = Column(Integer, primary_key=True)
    name = Column(String)

    def __repr__(self):
        return f"<Subject {self.id} {self.name}>"

class SubjectResult(Base):
    __tablename__ = 'subject_results'
    id = Column(Integer, primary_key=True)
    student_id = Column(
        Integer,
        ForeignKey('students.id'),
        nullable=False)
    student = relationship('Student', back_populates='results')
    subject_id = Column(
        Integer,
        ForeignKey('subjects.id'),
        nullable=False)
    subject = relationship(Subject)
    score = Column(Integer)

    def __repr__(self):
        return f"<SubjectResult for student {self.student_id} in subject {self.subject_id}>"

Base.metadata.create_all(engine)
Session = sessionmaker(bind=engine)
dbsession = Session()

student_1 = Student(name='Bob')
student_2 = Student(name='Sally')
sub_1 = Subject(name='Math')
sub_2 = Subject(name='English')
dbsession.add_all([
    student_1,
    student_2,
    sub_1,
    sub_2,
    SubjectResult(student=student_1, subject=sub_1, score=80),
    SubjectResult(student=student_1, subject=sub_2, score=70),
    SubjectResult(student=student_2, subject=sub_1, score=60),
    SubjectResult(student=student_2, subject=sub_2, score=50)
])

dbsession.commit()
```
It seems to work OK when the relationship is being lazy loaded (?)
```
>>> students = dbsession.query(Student).all()
>>> [s.math_result for s in students]
[[<SubjectResult for student 1 in subject 1>], [<SubjectResult for student 2 in subject 1>]]
```

But when I try to eager load it for multiple Students I end up with unexpected results
```
>>> students = dbsession.query(Student).options(selectinload("math_result")).all()
>>> [s.math_result for s in students]
[[<SubjectResult for student 1 in subject 1>, <SubjectResult for student 2 in subject 1>], [<SubjectResult for student 1 in subject 1>, <SubjectResult for student 2 in subject 1>]]
```

Is this even the right approach to get this working? I had a read of the docs and [Specifying Alternate Join Conditions](https://docs.sqlalchemy.org/en/14/orm/join_conditions.html#specifying-alternate-join-conditions)
 / [Composite 'Secondary' Joins](https://docs.sqlalchemy.org/en/14/orm/join_conditions.html#composite-secondary-joins ) seem similar but I couldn't quite grok how to transform those examples into what I'm trying to do.

I've also had a go getting it working as a [query enabled property](https://docs.sqlalchemy.org/en/14/orm/join_conditions.html#building-query-enabled-properties) but that seemed even worse with a SELECT being issued every time the property was accessed?

**Comment by zzzeek** [comment, marked as answer]
hey there -

this is close, but I think dont use secondary and instead use [relationship to aliased class](https://docs.sqlalchemy.org/en/14/orm/join_conditions.html#relationship-to-aliased-class).  this seems to satisfy the criteria: "we seek to join from A to B, making use of any number of C, D, etc. in between, however there are also join conditions between A and B directly. "

if you put that join() into an `aliased(SubjectResult)` object as in the example documentation and then do a primaryjoin to that aliased(), that should do it.  put viewonly=True on it too.

**Reply by Basskip** [reply]
That seems to work a treat, thanks!

Does it have to be viewonly=True? It would be convenient to be able to change values on the SubjectResult, but I see the docs also warn about doing so being error prone.

**Reply by zzzeek** [reply]
viewonly on Student.math_result only means that that one particular collection won't persist changes to it.  if you want to associate StudentResult objects with Student you append them to Student.results.    you aren't going to want to say Student.math_result = some_result because that doesn't validate that `some_result` is in fact associated with Math
